### ISSUE TITLE
Tracking issue: defaulting to Meson - follow-up tasks

### ISSUE THREAD

**Issue by rgommers** [numpy/numpy#23981]
After gh-23838 we now default to building with Meson (via `meson-python`, when invoking a build with `pip install .` or similar) in `main`. There are a number of more niche build configs not yet ported, and a few things were `xfail`-ed and need fixing up. Here is a list, roughly in order of priority:

- [x] Build nightlies for Python 3.12 to unblock other projects from testing with 3.12 (xref gh-23991)
- [x] Improve how CBLAS and ILP64 flags/dependencies are handled (xref gh-23909, and also automate passing `CFLAGS`/`CXXFLAGS` - I'll work on this next): see gh-23984
- [x] Finish implementing all BLAS/LAPACK auto-selection and control options (e.g. the equivalent of `NPY_BLAS_ORDER`). Done in gh-24893
- [x] Fix up the issue with PyPy and re-enable the `pypy3.9` CI job on Azure. Done in gh-24238.
- [x] Integrate SIMD support (the relevant CI jobs are still going through `runtests.py`)
- [x] Update the build docs (partially done in gh-24573; the rest in gh-25350).
- [x] Fix and re-eanble the 32-bit Windows wheels builds with MSVC on Azure (or move to GHA, or drop? - see gh-23717). Re-introducing on GHA in gh-24279.
- [x] Fix the `long double` representation check so the builds works on some of the more niche platforms again (xref gh-23972)
- [x] Fix the `xfail`-ed [`test_cython`](https://github.com/numpy/numpy/blob/11a3cef096ec3ab346cb0c7c43d1878d6174da72/numpy/random/tests/test_extending.py#L49) for Windows due to a missing `npyrandom.lib` (done in gh-23983)
- [x] Move Travis CI jobs over to Meson
- [x] Check that all CI jobs have been moved over (e.g., by removing `runtests.py` and `setup.py`, then running CI and ensuring nothing fails). Done - there were a lot left. See gh-24410 for the tracking issue for CI migration.
- [x] Investigate how to fix the `bin/f2py3X` CLI install, in addition to the `f2py` and `f2py3` ones (see project.scripts` in `pyproject.toml`, and commit f22a33b71). Or can/should that interface be dropped? Multiple installs into a single env is considered bad practice by now. Dropped in gh-24235.
- [ ] Investigate and fix the two `einsum` test failures with `int8`/`uint8` that were xfailed in gh-23838 (EDIT: these two tests fail with `clang-cl` on Windows too, see https://github.com/numpy/numpy/pull/20866#issuecomment-1631346830). EDIT: opened separate issue: gh-24732
- [x] Fix the array-api entrypoint installation (see `xfail` in `numpy/tests/test_public_api.py`). Done in gh-24249.
- [x] Fix the `longdouble` string repr test failures on `aarch64` - see gh-23974 and gh-23985
- [x] Move the Pyodide CI job over to Meson
- [x] Clean up the apparently unused `skip` statements under `[tool.cibuildwheel]` in `pyproject.toml`; the platforms to build for are controlled in `.github/workflows/wheels.yml`
- [x] Compare all files included in sdist and wheel between the setup.py and Meson builds, ensure nothing went missing
- [x] Fix `numpy.__version__` (see comment further down)
- [ ] Remove usages of `numpy.distutils` in user-facing docs. [Here](https://numpy.org/doc/stable/user/c-info.ufunc-tutorial.html) is one example.

**Comment by charris** [comment]
Note that cibuildwheel now supports Python 3.12.0b2.

**Comment by mattip** [comment]
`setup.py` and `runtests.py` (which uses `setup.py` underneath) are still used in the `tools/travis-test.sh` script.

**Comment by rgommers** [comment]
> `setup.py` and `runtests.py` (which uses `setup.py` underneath) are still used in the `tools/travis-test.sh` script.

Thanks, included two action items related to that, one for Travis CI and one to ensure nothing else was forgotten in any CI jobs.

**Comment by rgommers** [comment]
> Fix the `xfail`-ed [`test_cython`](https://github.com/numpy/numpy/blob/11a3cef096ec3ab346cb0c7c43d1878d6174da72/numpy/random/tests/test_extending.py#L49) for Windows due to a missing `npyrandom.lib`

This is due to the produced static library being named `libnpyrandom.a` rather than `libnpyrandom.lib`, and `distutils` (used by `test_cython`) not understanding that. For why, see https://mesonbuild.com/FAQ.html#why-does-building-my-project-with-msvc-output-static-libraries-called-libfooa. 

We're planning to drop this static library completely in 2.0, and if we don't get to that we can keep the `.a` file extension I'd say. For 1.26.0 though, we're better off using `.lib` in order to not break SciPy's `setup.py` Windows build (that may still have to work with 1.26.0).

**Comment by ngoldbaum** [comment]
It would be nice if the version information from versioneer could percolate up to meson at install time, like the old setuptools build. This is a little confusing:

```
$ pip install .
< a bunch of output elided >
Successfully installed numpy-2.0.0.dev0
$ python -c "import numpy; print(numpy.__version__)"
1.25.0rc1+265.g619a30d9a
```

**Comment by rgommers** [comment]
The exact same thing (including git hash in wheel filename and in `.dist-info/METADATA`) is not really possible, there's no mechanism for propagating such dynamic info back up to `meson-python`. `__version__` needs fixing though, it should be `2.0.0.dev0+xxx`. I added a task to the issue description.

However, I'm also (yet again) wondering what on earth `versioneer` is doing here. The most recent tag is `v1.25.0`, but in the root of a clean repo:
```python
>>> import versioneer
>>> versioneer.get_version()
'1.25.0rc1+264.gd38fc02fa'
```
Each time I look at `versioneer` I get a little annoyed, because it's ~3,000 lines of essentially garbage incomprehensible code split across `versioneer.py` and `numpy/_version.py` for doing exactly the same job as the ~120 LoC in https://github.com/scipy/scipy/blob/main/tools/version_utils.py.

I vote we get rid of it, and do exactly what SciPy does here - both lowest effort and a good cleanup.

**Comment by seberg** [comment]
RAPIDS is also going away from versioneer use as I understand it and I think many projects are.  Happy if we do also.

**Comment by charris** [comment]
There is a missing tag.

**Comment by charris** [comment]
I think some funny merges have taken place, messing up the history.

**Comment by charris** [comment]
And I think I know what happened, investigating.

**Comment by ngoldbaum** [comment]
For the docs, I think a reasonable editable development build command we can suggest (following [pandas](https://pandas.pydata.org/docs/dev/development/contributing_environment.html#step-3-build-and-install-pandas)):

```
python -m pip install -ve . --no-build-isolation --config-settings editable-verbose=true
```

In addition, we should document how to pass arguments via meson-python to meson, e.g. here's how to create a debug build and override the default to produce a release build:

```
python -m pip install . --config-settings setup-args="-Dbuildtype=debug"
```

Just commenting here with that for now because it took 15 minutes or so to figure out the proper way to pass arguments from pip into meson-python and then into meson.

**Comment by rgommers** [comment]
Good points @ngoldbaum. Do you find the `editable-verbose=true` useful/necessary? I don't have much experience with it yet; happy with either choice there.

Also, to save a bit of typing, `--config-settings` can be abbreviated now with an up-to-date `pip` to `-C`:
```
python -m pip install -ve . --no-build-isolation -Ceditable-verbose=true
python -m pip install . -Csetup-args="-Dbuildtype=debug"
```

**Comment by ngoldbaum** [comment]
> Do you find the editable-verbose=true useful/necessary?

When pandas made this change several contributors were confused by the way builds can happen on import but with no terminal output, so they added the `editable-verbose` bit to the build instructions to mitigate that.

**Comment by ngoldbaum** [comment]
Small addendum for `editable-verbose`, it may be painful for numpy to suggest that in general, because lots and lots of scripts out there are parsing the output of `import numpy as np; print(np.get_include())`. That snippet suddenly has extra output if you set `editable-verbose=True`. It's not hard to fix that but it's going to cause a lot of confusing compilation issues in other projects that depend on numpy if someone builds against an editable build of numpy.

**Comment by rgommers** [comment]
Good point, best to stay with the default behavior then. I think it's a matter of learning what's up with the delay on import. It's still significantly better than the distutils/setuptools behavior, where you get no auto-rebuild but simply a broken package that you then manually have to rebuild with `pip install -e .` or `python setup.py develop`. So some docs should be adequate here I'd think.

**Comment by charris** [comment]
I just deleted the v1.25.0rc1 tag, which seems the simplest solution. We don't really need it anymore, and unless you pulled it the problem should be fixed. It is ugly, but if no one sees it I'm happy (whistles innocently).

**Comment by rgommers** [comment]
The `__version__` issue seems fixed now; I get `2.0.0.dev0+279.gdbc454a15` on `main`. Thanks for untangling that Chuck.

**Comment by charris** [comment]
> The __version__ issue seems fixed now;

One thing I learned was that reverting a merge undoes the data from the merged branch, but the link is still there. At least, that is what I gathered when looking for info. How git implements that, I have no idea. But it seems that tags aren't data.

**Comment by eli-schwartz** [comment]
All commits depend on the full commit history of what came before it. Reverting a commit just means applying the opposite change to return to the previous codebase -- git still records that you made the change, and then undid the change.

Tags are also data, but they're the same type of data as a branch -- they depend on a commit, but other commits don't depend on *them*. Just like you can delete a branch (and all the commits exclusive to that branch) so too you can delete a tag (and all the commits exclusive to that tag).

**Comment by ngoldbaum** [comment]
I'm not sure if this is peculiar to my setup, but it looks like editable builds aren't ending up with a `_numpyconfig.h` file in the public include directory, `/numpy/core/include/numpy`. I only noticed today after doing a `git clean -fxd`, so others may not have run into this yet unless they start from a fresh clone of numpy. I'm not sure how to debug meson to figure out why it's not getting installed, the `meson.build` entry for it indicates it should be getting installed:

https://github.com/numpy/numpy/blob/c6e1fd039da2e6127ba0798f327e9e1b122d7534/numpy/core/meson.build#L490-L496

The missing file breaks the pandas build:

```
  ../../../numpy/numpy/core/include/numpy/numpyconfig.h:4:10: fatal error: _numpyconfig.h: No such file or directory
      4 | #include "_numpyconfig.h"
        |          ^~~~~~~~~~~~~~~~
  compilation terminated.
```

**Comment by ngoldbaum** [comment]
It turns out this is the expected behavior for editable builds with meson-python. This is something that should likely be highlighted in documentation as I'm sure I'm not the only one who tries to build other projects against an editable build of numpy.

**Comment by eli-schwartz** [comment]
If I understand the problem correctly, it is that editable installs read the data directly from the source/build tree together with an importlib loader that remaps files inside the python import system to ensure they can be imported correctly -- no promise is made regarding its usability for purposes other than importing code, I think?

The includedirs exposed by numpy in an editable install would have to pass the appropriate -I paths for both the source tree and the build tree, so that the compiler can detect files from a merged overall layout. This assumes that one file including another file uses the compiler search-path friendly form, rather than an include which is relative to the file the preprocessor is currently processing.

**Comment by rgommers** [comment]
I think you can add `meson.build` to `py.install_sources` in the relevant `meson.build` file in order to install it. That is how it works for any Python/data/pxd/etc file that is to be installed.

**Comment by ngoldbaum** [comment]
> The includedirs exposed by numpy in an editable install would have to pass the appropriate -I paths for both the source tree and the build tree, so that the compiler can detect files from a merged overall layout. 

Ah good point, the headers I need *are* included in the build directory:

```
$ ls cp310d/numpy/core/*.h
cp310d/numpy/core/config.h            cp310d/numpy/core/npy_math_internal.h  cp310d/numpy/core/__ufunc_api.h
cp310d/numpy/core/__multiarray_api.h  cp310d/numpy/core/_numpyconfig.h       cp310d/numpy/core/_umath_doc_generated.h
```

One wrinkle is that due to the way the numpy build is set up, the headers end up one level up from the include directory.

Is there a way to detect at runtime whether numpy was built as an editable install? If so we should be able to include the appropriate path in the build directory in `numpy.get_include()`.

**Comment by mattip** [comment]
If we cannot fix the way the numpy build is set up, I think `numpy.get_include()` could try the fast-path option, and fall back if it is not found.

**Comment by rgommers** [comment]
> Is there a way to detect at runtime whether numpy was built as an editable install? 

No supported way, but I suspect that checking if `np.__path__` is an empty list instead of a path will be a decent heuristic.

> If so we should be able to include the appropriate path in the build directory in `numpy.get_include()`.

There's so many ways for this to break, and it's not really supposed to work: https://peps.python.org/pep-0660/#limitations. We have Python, C and Cython APIs, and only the Python ones are guaranteed to work. The `.pxd` files we carry are all non-generated and I suspect simple enough that they will work, but I haven't tried. For C, you could try to quickly hack `get_include` and see if that solves your problem. However, as @eli-schwartz points out above, our include style will be a problem. We use quotes style for numpy-internal includes, e.g. `#include "numpy/npy_math.h"`. Changing all that would be a lot of churn.

**Comment by seberg** [comment]
> Is there a way to detect at runtime whether numpy was built as an editable install?

Couldn't you add a `_distributor_init.py` to monkey-patch this?  In a sense an inplace install does seem like a custom distribution (at least I don't see it clashing with the main purpose of that file).

**Comment by stefanv** [comment]
FWIW the versioneer issue again came up during the SciPy sprints, where it created a lot of confusion that we kept getting `1.25.0rc1+517.gbd7703dbd`. After a lot of wasted time, it looks like it grabs tags chronologically, so that we need a new `2.0.0dev1` tag on `main`.

I agree with @rgommers above that versioneer only complicates version discovery. I'll make a PR which proposes to adopt the scipy or skimage approach.

**Comment by stefanv** [comment]
The tags need to be pulled down with the `--prune` flag, otherwise they don't get removed. But even though the issue is now "fixed", it doesn't affect my position on versioneer :)

**Comment by charris** [comment]
> it doesn't affect my position on versioneer :)

I'm fine with that, especially if someone else does the work :) It should be easy to do, the big decision is whether or not is should be tag based, `git describe` does most of the work if we keep the tags. I'm also not opposed to adding a `v2.0.0.dev1` tag.

**Comment by rgommers** [comment]
Better not to have it tag-based, that only has downsides as far as I can tell.

**Comment by melissawm** [comment]
For the docs:

1. Would it be reasonable to recommend `spin` from 1.26 on?
2. Do we want to remove or have both instructions (spin/meson and distutils stuff) for now?
3. When we update both https://numpy.org/devdocs/user/building.html and https://numpy.org/devdocs/dev/development_environment.html, do we want to focus on `pip install` for the "Building from source" page and have the more configurable steps in the "Setting up for development" page, or not really?
4. Besides moving the `building_with_meson.md` file into these pages, any other recommendations? The SciPy docs are pretty much in shape so we could also link out some stuff, for example the [Compiler selection and customizing a build page](http://scipy.github.io/devdocs/building/compilers_and_options.html)

**Comment by rgommers** [comment]
Good questions Melissa. 

- let's not update the docs in the 1.26.x docs beyond release notes - no one except a few maintainers who know what they are doing should be targeting that branch.
- for `main`/2.0, let's just switch over to meson/spin in one go. I don't think that it's useful to keep the old docs, we can simply add a note that if someone still really needs to build with setup.py (assuming we even keep that until 2.0) to look at the 1.25.x docs.
- Given that the build docs are now split into three places (as you pointed out above), I'd be inclined to remove that split and only keep https://numpy.org/devdocs/user/building.html. And replace the content with exactly what SciPy has now, and tweak it from there. Those docs are in much better shape, and more applicable than a 1:1 translation of the distutils-based docs.

**Comment by charris** [comment]
> for main/2.0, let's just switch over to meson/spin in one go.

Setup.py no longer works for me in the main branch, meson does fine. Unless we want to spend time fixing up setup.py, meson it is.

**Comment by rgommers** [comment]
I checked sdist contents, there are some differences but they all look fine to me. To reproduce

Build sdists with:
```
python setup.py sdist  # then copy to ../tmp/distutils/ and unpack with `tar -xzvf`
python -m build --sdist  # then copy to ../tmp/meson/ and unpack with `tar -xzvf`
```

Then run this script:

<details>

```python
import os
from filecmp import dircmp

def print_diff_files(dcmp):
    for name in dcmp.left_list:
        if name not in dcmp.right_list:
            print(f"{name} missing from {dcmp.right}")
            
    for name in dcmp.right_list:
        if name not in dcmp.left_list:
            print(f"{name} missing from {dcmp.left}")

    for name in dcmp.diff_files:
        print("diff_file %s found in %s and %s" % (name, dcmp.left,
              dcmp.right))

    for sub_dcmp in dcmp.subdirs.values():
        print_diff_files(sub_dcmp)


meson_dir = 'meson/numpy-2.0.0.dev0'
distutils_dir = 'distutils/numpy-2.0.0.dev0+661.g35d947160'
dcmp = dircmp(distutils_dir, meson_dir)
print_diff_files(dcmp) 
```

</details>

That gives:
```
.circleci missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.cirrus.star missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.clang-format missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.codecov.yml missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.devcontainer missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.gitattributes missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.github missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.gitignore missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.gitmodules missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.lgtm.yml missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.mailmap missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.spin missing from distutils/numpy-2.0.0.dev0+661.g35d947160
.travis.yml missing from distutils/numpy-2.0.0.dev0+661.g35d947160
LICENSES_bundled.txt missing from distutils/numpy-2.0.0.dev0+661.g35d947160
_version_meson.py missing from distutils/numpy-2.0.0.dev0+661.g35d947160
azure-pipelines.yml missing from distutils/numpy-2.0.0.dev0+661.g35d947160
azure-steps-windows.yml missing from distutils/numpy-2.0.0.dev0+661.g35d947160
branding missing from distutils/numpy-2.0.0.dev0+661.g35d947160
diff_file LICENSE.txt found in distutils/numpy-2.0.0.dev0+661.g35d947160 and meson/numpy-2.0.0.dev0
diff_file PKG-INFO found in distutils/numpy-2.0.0.dev0+661.g35d947160 and meson/numpy-2.0.0.dev0
diff_file _version.py found in distutils/numpy-2.0.0.dev0+661.g35d947160/numpy and meson/numpy-2.0.0.dev0/nump
```

So the meson-generated sdist has more files, those are the ones that are explicitly excluded by `distutils` in `MANIFEST.in`. Given that the sdist is supposed to be a snapshot of the release and those files hardly take up any space (10.90 MB vs. 10.83 MB), I'd say we should not exclude them. 

For the files with differences:
- the `LICENSE.txt` one is because we concatenate `LICENSES_BUNDLED.txt` into it in `setup.py`, but in the meson build we don't (it's done in the CI jobs for wheel builds instead).
- the `PKG-INFO` difference is because the metadata comes from `pyproject.toml` rather than `setup.py`, and those diverges (changes look fine)
- the `numpy/_version.py` one is because of `versioneer` behavior, which is in the process of being fixed in gh-24196

tl;dr all good for sdist content.

**Comment by stefanv** [comment]
For those who need to modify sdist content, that happens in `.gitattributes`.
This uses/mirrors the [`git-archive`](https://git-scm.com/docs/git-archive#ATTRIBUTES) command.

**Comment by eli-schwartz** [comment]
> So the meson-generated sdist has more files, those are the ones that are explicitly excluded by `distutils` in `MANIFEST.in`. Given that the sdist is supposed to be a snapshot of the release and those files hardly take up any space (10.90 MB vs. 10.83 MB), I'd say we should not exclude them.

They could still be excluded with meson, just using `.gitattributes` via `export-ignore`. And those files are, by and large, only useful when specifically working with a git repo (due to containing git repository CI triggers) -- so I would say they don't really count as a "snapshot of the release". :)

.spin at least seems pretty useful for a release though. Probably so does .clang-format.

**Comment by stefanv** [comment]
For scikit-image, we ignore all dot-files, and then include back the ones we want:

```
$ cat .gitattributes
.* export-ignore
.spin -export-ignore
azure-pipelines.yml export-ignore
```

**Comment by rgommers** [comment]
Same for wheels, comparing `python -m build --wheel` with `python setup.py bdist_wheel`:

Script:

<details>

```
import os
from filecmp import dircmp

def print_diff_files(dcmp):
    for name in dcmp.left_list:
        if name not in dcmp.right_list:
            print(f"{name} missing from {dcmp.right}")
            
    for name in dcmp.right_list:
        if name not in dcmp.left_list:
            print(f"{name} missing from {dcmp.left}")

    for name in dcmp.diff_files:
        print("diff_file %s found in %s and %s" % (name, dcmp.left,
              dcmp.right))

    for sub_dcmp in dcmp.subdirs.values():
        print_diff_files(sub_dcmp)


meson_sdist_dir = 'meson/numpy-2.0.0.dev0'
distutils_sdist_dir = 'distutils/numpy-2.0.0.dev0+661.g35d947160'
dcmp = dircmp(distutils_sdist_dir, meson_sdist_dir)
print('DIFFERENCES FOR SDIST\n')
print_diff_files(dcmp) 
print('\n\n')

def print_diff_tree(dcmp, ):
    print(f"Missing from {dcmp.right}:")
    missing_names = []
    for name in dcmp.left_list:
        if name not in dcmp.right_list:
            missing_names.append(os.path.join(dcmp.right, name))

    for name in missing_names:
        print(name)


meson_wheel_dir = 'meson-wheel/numpy'
distutils_wheel_dir = 'distutils-wheel/numpy'
dcmp = dircmp(distutils_wheel_dir, meson_wheel_dir)
print('DIFFERENCES FOR WHEEL\n')
print_diff_tree(dcmp) 
print('\n\n')
dcmp = dircmp(meson_wheel_dir, distutils_wheel_dir)
print_diff_tree(dcmp) 
```

</details>

```
Missing from meson-wheel/numpy:
meson-wheel/numpy/LICENSE.txt
meson-wheel/numpy/setup.py

Missing from distutils-wheel/numpy:
distutils-wheel/numpy/_version_meson.py
```
That all looks as expected. The Meson wheel is missing two files from the `.dist-info` directory:
```
LICENSES_bundled.txt
top_level.txt
```
I think `top_level.txt` is setuptools-specific, and https://discuss.python.org/t/record-the-top-level-names-of-a-wheel-in-metadata/29494/37?page=2 says it doesn't need to exist I think. `LICENSES_BUNDLED.txt` is missing and should be added.

**Comment by rgommers** [comment]
Sure, I know how to exclude files, I just don't really see the point of doing so. My experience with `MANIFEST.in` is that it's only a source of bugs, inevitable some config files end up included and some excluded. Given that the sdist is simply a tarball, why would we not make it a complete snapshot of the repo? Less work and no downside I think.

**Comment by rgommers** [comment]
> `LICENSES_BUNDLED.txt` is missing and should be added.

That I knew about, says the comment in `pyproject.toml`. It's pending the arrival of PEP 639, which is taking an eternity. There's no issue for the wheels we release, because the two files get appended into `LICENSE.txt`. So we should be all good here.

**Comment by rgommers** [comment]
I sent a summary of the current status to the mailing list: https://mail.python.org/archives/list/numpy-discussion@python.org/thread/NQZE6SPXFJGXP5ZDXCBCCOTM2JDV2HHV/

**Comment by charris** [comment]
There are still a few unchecked boxes, so pushing off to 1.26.1. I'm most concerned about the einsum failures using clang.

**Comment by rgommers** [comment]
> I'm most concerned about the einsum failures using clang.

I've opened gh-24732 as a separate issue for that. It is a lot lower-prio to me than the other two items, and I don't have the hardware needed to reproduce the problem, so I'll leave that to whoever feels like digging in.

**Comment by ngoldbaum** [comment]
I just noticed there are still some spots in the docs that tell users to use `numpy.distutils` so I added a task to the issue to deal with those.

**Comment by jorisvandenbossche** [comment]
I was helping a colleague to install a numpy debug build from source yesterday, and we ran into a few issues / confusions with the new meson / spin setup. Summarizing our experience here, in case that has some useful pointers for where the docs can be improved:

* I had an old numpy clone that I updated, so as a starter I ran into https://github.com/numpy/numpy/issues/25132
* We started from https://numpy.org/devdocs/user/building.html#building-from-source (which was pointed to from https://numpy.org/devdocs/dev/development_environment.html#building-from-source). The basic `pip install` instructions there work fine to get an install, except that our goal was a debug build.  But it has a note on "Advanced building with Meson", which points to the SciPy docs for more information (https://docs.scipy.org/doc/scipy/building/compilers_and_options.html). However, I think those instructions from SciPy might no longer be valid for numpy? (for example my understanding is that the `meson setup` doesn't work right now for numpy because you vendor meson?)
  And so the `meson setup build --buildtype debug` mentioned there didn't work to get a debug build through `pip install`
* Going back to the original doc page, there is a section about debugging (https://numpy.org/devdocs/dev/development_environment.html#debugging), which mentions how to build numpy in debug mode using `spin`. First note here is that this full page doesn't mention what `spin` actually is. It might be explained elsewhere in the docs, but this page is the first one (after git) in the "Development" section of the docs. For example mentioning that `spin` is something you need to install separately (`pip install spin`) might be useful (it could also have been a tool included in the numpy repo itself, or similar to a makefile)
* We succeeded to get a debug build with the `spin build -- -Dbuildtype=debug ..` command mentioned there, but what was unclear to me is what this exactly does. Did it only _build_ numpy, or did it also _install_ it in my dev env? My understanding now is that this essentially only builds numpy and "installs" it in a specific directory, but not in your environment, so you need to specifically add this directory on the path to use this build of numpy. This is what `spin test ..`, `spin gdb ..` etc do for you (from seeing the console output doing an `export PYTHONPATH= ..`). This is fine if you use those tools, but in our case we wanted to install numpy into our pyarrow dev environment (copying that `export` command manually worked to get pyarrow rebuilt). It's not clear to me if `spin` also has the option to do an actual install into your environment.
* Using `spin gdb` worked fine for me, my colleague had to use `lldb`. But that's without `spin`, so that means you need to manually set the python path to pick up the numpy build from `spin build`? (which is not mentioned in the docs)

In the end, once we got the gdb running with the numpy debug build, it pointed us directly to the information we needed to know, so a good ending to the story ;)

I assume that for our case, it might have been easiest to just stick with the initial `pip install` we tried, and I noticed that higher up here in this issue, it is mentioned how you can pass the debug build flag through pip (https://github.com/numpy/numpy/issues/23981#issuecomment-1597625463) with `python -m pip install . --config-settings setup-args="-Dbuildtype=debug"`. I agree that it would be nice to document this more clearly.

**Comment by rgommers** [comment]
Thanks for sharing your experience @jorisvandenbossche. Sorry it was painful - I guess we should update these docs asap.

> This is fine if you use those tools, but in our case we wanted to install numpy into our pyarrow dev environment (copying that `export` command manually worked to get pyarrow rebuilt). It's not clear to me if `spin` also has the option to do an actual install into your environment.

`spin` doesn't have that option, it is _only_ useful and meant for working on `numpy` itself. We need to clearly separate "development work on numpy" from "building and installing numpy".

> I assume that for our case, it might have been easiest to just stick with the initial `pip install` we tried, and I noticed that higher up here in this issue, it is mentioned how you can pass the debug build flag through pip ([#23981 (comment)](https://github.com/numpy/numpy/issues/23981#issuecomment-1597625463)) with `python -m pip install . --config-settings setup-args="-Dbuildtype=debug"`. I agree that it would be nice to document this more clearly.

Yes indeed, that should be all that is needed here. And with an up-to-date version of `pip`, it gets even shorted (the `-C` alias for `--config-settings` was added): 
```
pip install . -Csetup-args=-Dbuildtype=debug
```

**Comment by stefanv** [comment]
Wrt potential spin improvements:

- We just released 0.8 with an lldb command
- We added an install command, via pip. I can update its documentation to show how to pass in the debug flag.

**Comment by rgommers** [comment]
This is basically done and the tracking issue has served its purpose. I'm sure there's little more loose ends somewhere, but those should get their own issues by now.

> Remove usages of `numpy.distutils` in user-facing docs. [Here](https://numpy.org/doc/stable/user/c-info.ufunc-tutorial.html) is one example.

I'll open a separate issue for this, that's a doc task that is more related to the deprecation of `numpy.distutils` than to the introduction of Meson - it's just actionable now.
